features/0434-outofband/README.md

```diff
@@ -0,0 +1,417 @@
+
+The agent that generates the out of band message and makes it available to the other party.
+
+#### receiver
+
+The agent that receives the out of band message and decides how to respond. There is no out of band protocol message with which the receiver will respond. Rather, if they respond, they will use a message from another protocol that the sender will understand.
+
+### States
+
+The state machines for the sender and receiver are a bit odd for the out of band protocol because of it being a single message protocol with a response from a different protocol. In the following state machine diagrams we generically describe the response message from the *sender* as being a DIDComm message without associating the message with a specific protocol.
+
+The *sender* state machine is as follows:
+
+[![sender state machine google doc](state-machine-sender.png)](https://docs.google.com/spreadsheets/d/1o1szGCMx6FzrxfAKV76IEYmeKRJqvV3gd1r3Qfv1Q9o/edit#gid=1176419697)
+
+Note the "optional" reference under the second event in the `await-response` state. That is to indicate that an out of band message might be a single use message with a transition to done, or reusable message (received by many receivers) with a transition back to `await-response`.
+
+The *receiver* state machine is as follows:
+
+[![receiver state machine google doc](state-machine-receiver.png)](https://docs.google.com/spreadsheets/d/1o1szGCMx6FzrxfAKV76IEYmeKRJqvV3gd1r3Qfv1Q9o/edit#gid=10874521)
+
+Worth noting is the first event of the `done` state, where the receiver may receive the message multiple times. This represents, for example, an agent returning to the same website and being greeted with the same QR code each time.
+
+### Messages
+
+There are two messages in the out of band protocol. They are both sent by the *sender*. Only one of the message types is ever used in an instance of the protocol.
+
+#### Message Type: `https://didcomm.org/outofband/VER/invitation`
+
+```jsonc
+{
+  "@type": "https://didcomm.org/outofband/VER/invitation",
+  "@id": "<id used for context as pthid>",
+  "label": "Faber College",
+  "protocols": [
+      "https://didcomm.org/did-exchange/1.0",
+      "https://didcomm.org/connections/1.0"
+      ]
+  "service": ["did:sov:LjgpST2rjsoxYegQDRm7EL"]
+}
+```
+
+The items in the message are:
+
+- `@type` - the DIDComm message type
+- `@id` - the unique ID of the message. The ID should be used as the **parent** thread ID (`pthid`) for the response message, rather than the more common thread ID (`thid`) of the response message. This enables multiple uses of a single `invitation` message.
+- `label` - a self-attested string that the receiver may want to display to the user.
+- `protocols` - array of protocols in the order of preference of the sender that the receiver can use in responding to the message. As implied by the message type of invitation, these are not arbitrary protocols but rather protocols that result in the establishment of a connection.
```

Comment:
> in the order of preference

Cannot rely on ordering across all JSON libraries. Suggest we punt this priority ordering?

Reply:
Good point. I think I'm OK with dropping the ordering/priority requirement, as the community update process should allow for the deprecating protocols.

Others have a comment on this?  @andrewwhitehead @TelegramSam @KitHat ?

Reply:
I didn't see a copy later in the RFC, but assume we would similarly drop the ordering of references related to the DIDComm service entries?

Reply:
The priority order is only a suggestion to the receiver on which protocol or service to choose. In the unlikely case where the elements are reordered there shouldn't be any negative impact.

That said, it might be an improvement to allow an object representation for protocols as well as the string representation, so that additional properties can be supported in future. Maybe just `{"name": "<protocol name>"}`.

Reply:
Re. priority order: this is one of the obnoxious things about JSON-LD, IMO: it explicitly contradicts ordering rules in JSON. JSON parsers are required to maintain order of JSON arrays, or they're not JSON-compliant. So I don't actually think George's concern is likely to occur. Do we know for a fact that it does? Having said that, a JSON-LD parse that reads DIDComm messages would probably make the very mistake he predicts, so maybe it's still an issue.

On a different subject: I don't think the invitation with protocol names is enough info. We need an invitation to specify both the name of the protocol, and the name of the role in the protocol that the inviter proposes the invitee to take. In both of the protocols we're naming here, the role implied for the invitee is `requester`, but there's no guarantee that future connection-like protocols will work with that implicit assumption.

Reply:
My vote is to leave the text about the ordering as in the unlikely event there is a re-ordering, there is no harm done, since any are by definition OK to use.

I disagree with the need for the role name to be included.  By definition, the recipient will be the initiator of whatever protocol is selected.  Further, an "invitation" is by definition being used to create a connection, so there is little need to specify the role. It would also increase the size of the invitation, making URL shortening more likely to be required.

From the others, what about Andrew's suggestion about `{"name": "<protocol name>"}`.

Reply:
I disagree with the analysis about the role name; see my longer comment further down. Specifically, I don't agree that an invitation should by definition be viewed as being used to create a connection. Spammers would love that. We need to make the goal a lot more explicit than just "hey, let's connect."

Reply:
I agree that JSON preserves order and 'higher-level' parsers that don't respect that are flawed.

Reply:
> there is no harm done, since any are by definition OK to use.

If a protocol is supported for legacy purposes, but really shouldn't be the default then this could present an edge case where it causes a problem.